compiler/rustc_borrowck/src/borrowck_errors.rs

```diff
@@ -233,17 +243,11 @@ impl<'cx, 'tcx> crate::MirBorrowckCtxt<'cx, 'tcx> {
         borrow_span: Span,
         desc: &str,
     ) -> DiagnosticBuilder<'cx, ErrorGuaranteed> {
-        let mut err = struct_span_err!(
-            self,
-            span,
-            E0506,
-            "cannot assign to {} because it is borrowed",
-            desc,
-        );
-
-        err.span_label(borrow_span, format!("borrow of {} occurs here", desc));
-        err.span_label(span, format!("assignment to borrowed {} occurs here", desc));
-        err
+        let err = AssignBorrowErr { desc, span };
+        let mut diag = self.infcx.tcx.sess.create_err(err);
+        diag.subdiagnostic(AssignBorrowLabel::Here { desc, borrow_span });
+        diag.subdiagnostic(AssignBorrowLabel::AssignHere { desc, span });
```

Comment:
If `AssignBorrowLabel` is only used with `AssignBorrowErr` then it can just be `#[label(..)]` attributes in the type. Alternatively, if `AssignBorrowErr` isn't used elsewhere and always has these labels, then they can be fields of the type and annotated with `#[subdiagnostic]`.